File: web/src/api/mtg.ts
import { api } from "./client";
import type { MtgDeckImportResult, MtgDeckStatusOut, MtgState, MtgZone } from "../types";

export function importMtgDeck(lobbyId: number, decklist: string) {
  return api.post<MtgDeckImportResult>(`/mtg/${lobbyId}/deck`, { decklist }).then((r) => r.data);
}

export function getMtgDeckStatus(lobbyId: number) {
  return api.get<MtgDeckStatusOut>(`/mtg/${lobbyId}/deck-status`).then((r) => r.data);
}

export function createMtgSession(lobbyId: number) {
  return api.post<MtgState>(`/mtg/${lobbyId}/session`).then((r) => r.data);
}

export function getMtgState(lobbyId: number) {
  return api.get<MtgState>(`/mtg/${lobbyId}/state`).then((r) => r.data);
}

export function drawMtgCard(lobbyId: number) {
  return api.post<MtgState>(`/mtg/${lobbyId}/draw`).then((r) => r.data);
}

export function shuffleMtgLibrary(lobbyId: number) {
  return api.post<MtgState>(`/mtg/${lobbyId}/shuffle`).then((r) => r.data);
}

export function moveMtgCard(
  lobbyId: number,
  params: {
    instanceId: string;
    ownerUserId: number;
    fromZone: MtgZone;
    toZone: MtgZone;
    x?: number;
    y?: number;
    // Only meaningful for hand -> battlefield (summoning).
    faceDown?: boolean;
  },
) {
  return api
    .post<MtgState>(`/mtg/${lobbyId}/move`, {
      instance_id: params.instanceId,
      owner_user_id: params.ownerUserId,
      from_zone: params.fromZone,
      to_zone: params.toZone,
      x: params.x,
      y: params.y,
      face_down: params.faceDown ?? false,
    })
    .then((r) => r.data);
}

export function tapMtgCard(lobbyId: number, instanceId: string, ownerUserId: number, tapped: boolean) {
  return api
    .post<MtgState>(`/mtg/${lobbyId}/tap`, { instance_id: instanceId, owner_user_id: ownerUserId, tapped })
    .then((r) => r.data);
}

export function updateMtgCounter(
  lobbyId: number,
  instanceId: string,
  ownerUserId: number,
  counterType: string,
  delta: number,
) {
  return api
    .post<MtgState>(`/mtg/${lobbyId}/counter`, {
      instance_id: instanceId,
      owner_user_id: ownerUserId,
      counter_type: counterType,
      delta,
    })
    .then((r) => r.data);
}

export function adjustMtgLife(lobbyId: number, targetUserId: number, delta: number) {
  return api.post<MtgState>(`/mtg/${lobbyId}/life`, { target_user_id: targetUserId, delta }).then((r) => r.data);
}

export function advanceMtgPhase(lobbyId: number) {
  return api.post<MtgState>(`/mtg/${lobbyId}/phase`).then((r) => r.data);
}

export function concedeMtg(lobbyId: number) {
  return api.post<MtgState>(`/mtg/${lobbyId}/concede`).then((r) => r.data);
}

File: api/app/schemas/mtg.py
from pydantic import BaseModel


class MtgDeckImportIn(BaseModel):
    decklist: str


class MtgDeckImportOut(BaseModel):
    card_count: int
    unresolved_names: list[str]


class MtgCardOut(BaseModel):
    id: str
    # None for a card in an opponent's hand, or a face-down battlefield
    # permanent viewed by anyone but its owner — the client renders a
    # card-back placeholder rather than the real name/image.
    name: str | None = None
    image_url: str | None = None
    tapped: bool = False
    counters: dict[str, int] = {}
    x: float = 0.5
    y: float = 0.5
    # Battlefield-only — always sent (even to non-owners, alongside a
    # nulled-out name/image) so the client can tell "hidden because it's
    # face-down" apart from "hidden because it's someone else's hand".
    face_down: bool = False


class MtgPlayerStateOut(BaseModel):
    user_id: int
    life: int
    library_count: int
    hand: list[MtgCardOut]
    battlefield: list[MtgCardOut]
    graveyard: list[MtgCardOut]
    exile: list[MtgCardOut]


class MtgStateOut(BaseModel):
    lobby_id: int
    turn_number: int
    active_user_id: int
    phase: str
    status: str
    winner_user_id: int | None
    # The fixed reference seat battlefield (x, y) coordinates are stored
    # relative to — the client rotates the whole board 180° when the
    # viewer isn't this player, so each player always sees their own side
    # at the bottom and the opponent's at the top, regardless of who
    # actually placed a given card.
    player1_user_id: int
    players: list[MtgPlayerStateOut]


class MtgMoveIn(BaseModel):
    instance_id: str
    # Whose zone this card lives in — not necessarily the acting player (an
    # opponent's removal/bounce spell moves a card between *its owner's*
    # zones, honor-system trusted the same as everything else here).
    owner_user_id: int
    from_zone: str
    to_zone: str
    x: float | None = None
    y: float | None = None
    # Only meaningful for hand -> battlefield (summoning) — whether to play
    # the card face-down (hidden from everyone but its owner) or face-up.
    face_down: bool = False


class MtgTapIn(BaseModel):
    instance_id: str
    owner_user_id: int
    tapped: bool


class MtgCounterIn(BaseModel):
    instance_id: str
    owner_user_id: int
    counter_type: str
    delta: int


class MtgLifeIn(BaseModel):
    target_user_id: int
    delta: int


class MtgDeckStatusEntry(BaseModel):
    user_id: int
    card_count: int


class MtgDeckStatusOut(BaseModel):
    players: list[MtgDeckStatusEntry]
